### ISSUE TITLE
Type subscript runtime exception false positive

### ISSUE THREAD

**Issue by sbdchd** [microsoft/pylance-release#1145]
<h3> Environment data </h3>

-   Language Server version: 2021.4.0
-   OS and version: darwin x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: off

<h3> Expected behaviour </h3>

Shouldn't report that indexing type will generate a runtime exception since the type isn't evaluated at runtime

<h3> Actual behaviour </h3>

error:

```
Subscript for class "OrderedDict" will generate runtime exception; enclose type annotation in quotes
```

<h3> Logs </h3>

Python Language Server Log
```
[Info  - 6:34:52 PM] Pylance language server 2021.4.0 (pyright 2918da1d) starting
[Info  - 6:34:52 PM] Server root directory: ~/.vscode/extensions/ms-python.vscode-pylance-2021.4.0/dist
```

<h3> Code Snippet / Additional information </h3>

![Screen Shot 2021-04-09 at 10 50 37 PM](https://user-images.githubusercontent.com/7340772/114256151-01cfe000-9986-11eb-9dcd-6c8efb56a84b.png)

if we run the following snippet it doesn't generate a runtime exception

```python
from collections import OrderedDict


def main() -> None:
    x: OrderedDict[str, str] = OrderedDict()
    print(x)


main()
```

**Comment by erictraut** [comment]
It's very strange that it does not generate a runtime exception in the context of a variable type annotation, but it does when used in the context of a return type or parameter type annotation.

If I modify your code as follows and then run it in a version of Python < 3.9 (I tried 3.6), it generates a runtime exception on the return type annotation.

```python
def main() -> OrderedDict[str, str]:
    x: OrderedDict[str, str] = OrderedDict()
    print(x)
    return x
```

```
    def main() -> OrderedDict[str, str]:
TypeError: 'type' object is not subscriptable
```

**Comment by erictraut** [comment]
I've updated the logic to differentiate between cases where this will cause runtime exceptions (parameter annotations, return type annotations, type aliases, and other non-annotation expressions) and cases where it will not (PEP 526-style variable type annotations). This change will be in the next release of Pylance.

Thanks for reporting this issue.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.4.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202141-14-april-2021

**Comment by fiendish** [comment]
Can this issue be re-opened? Python >= 3.9 does not have any problem with return type annotations.
